### PR TITLE
add low- and high-level Concept-Based Memory Reasoner (CMR) implementation

### PR DESCRIPTION
**Pull request by daviddebot**
This PR adds CMR across both low-level and high-level model API, with documentation updates and unit tests.

### CODE REVIEW

File: torch_concepts/nn/modules/high/models/cmr.py [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+
+        return {
+            'concept_input': no_rec[:, concept_indices],
+            'concept_target': target[:, concept_indices],
+            'task_input': no_rec[:, task_indices],
+            'task_input_with_rec': with_rec[:, task_indices],
+            'task_target': target[:, task_indices],
+        }
+
+    def shared_step(self, batch, step):
```
Comment (gdefe) [inline torch_concepts/nn/modules/high/models/cmr.py]: is it possible to avoid overriding the lightning shared step? it is OK to instead override the model forward since each model can have their own.

I see the problem is the task predictor which need to be done twice. Is is possible to play with the inference query in the forward?

File: torch_concepts/nn/modules/loss.py [hunk 2]
```diff
@@ -214,4 +214,103 @@ def forward(self, input: torch.Tensor, target: torch.Tensor) -> torch.Tensor:
         c_loss = self.concept_loss(concept_input, concept_target)
         t_loss = self.task_loss(task_input, task_target)
         
-        return c_loss * self.concept_weight + t_loss * self.task_weight
\ No newline at end of file
+        return c_loss * self.concept_weight + t_loss * self.task_weight
+
+
+class CMRLoss(nn.Module):
```
Comment (gdefe) [inline torch_concepts/nn/modules/loss.py]: It would be great not to have a dedicated loss for every model. It would be great to decompose the CMR loss into modular pieces. I would expect you just need to create an additional piece for the recursion part. To combine the traditional loss with this additional loss, see the way we now enable losses to be summed in the example 
pytorch_concepts/examples/utilization/2.2_model/13_composite_loss.py

File: torch_concepts/nn/modules/low/encoders/selector.py [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+"""
+import numpy as np
+import torch
+import torch.nn.functional as F
+
+
+from ..base.layer import BaseConceptLayer
+
+
+class CategoricalSelector(BaseConceptLayer):
```
Comment (gdefe) [inline torch_concepts/nn/modules/low/encoders/selector.py:15]: could you standardize the layer name with the convention we have in the library?

<OperationType><InputType>To<OutputType>

we have a new guide: https://pytorch-concepts.readthedocs.io/en/latest/guides/contributing_layer.html

File: torch_concepts/nn/modules/low/encoders/selector.py [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+        *args: Additional positional arguments for linear layers in the selector.
+        **kwargs: Additional keyword arguments for linear layers in the selector.
+
+    References:
+        Debot et al. "Interpretable Concept-Based Memory Reasoning", NeurIPS 2024. https://arxiv.org/abs/2407.15527
+    """
+    def __init__(
+        self,
+        in_latent: int,
+        out_exogenous: int,  # nb_rules
```
Comment (gdefe) [inline torch_concepts/nn/modules/low/encoders/selector.py:44]: all PyC layers should follow the signature:

        out_concepts: Union[int, Annotations],
        in_concepts:  Union[int, Annotations] = None,
        in_embeddings: Union[int, Annotations] = None,

what is the role of "out_exogenous" in this layer? A PyC layer only accept out_concepts as standard output dimensions. See our embedding encoders in nn/dense_layers.py for layers producing embeddings.

Is this layer really necessary as a standalone layer?
could this become a simple sequential in the model?

File: torch_concepts/nn/modules/low/encoders/selector.py [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+        Args:
+            latent: Input latent of shape (batch_size, in_latent).
+
+        Returns:
+            torch.Tensor: Mixing probabilities of shape
+                         (batch_size, out_concepts, out_exogenous).
+        """
+        mixing_coeff = self.selector(latent)
+        return torch.softmax(mixing_coeff, dim=-1)  # [Batch x Task x Memory]
```
Comment (gdefe) [inline torch_concepts/nn/modules/low/encoders/selector.py:108]: If you decide to keep the layer, activations should not be part of it (assume produce logits). The user will combine layers with activations in the parametrization of mid-level CPDs.

File: torch_concepts/nn/modules/low/predictors/rule.py
Comment (gdefe) [inline torch_concepts/nn/modules/low/predictors/rule.py:1]: Also here: 

- If these are PyC layers, they should match the nomenclature, and the signature of both the init and the forward, and should not activate the output. 
- If these are torch layers (output is not concept), there is more flexibility. In this case, we recommend to internally use the MLP class that we provide, which is very flexible.

File: torch_concepts/nn/modules/high/models/cmr.py [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+        self.pgm = self._build_model()
+        self.setup_inference(inference, inference_kwargs, train_inference, train_inference_kwargs)
+
+    def default_query(self, ground_truth):
+        """Train both CMR task paths in one inference query."""
+        query = super().default_query(ground_truth)
+        query["tasks_with_rec"] = None
+        return query
+
+    def forward(self, query, evidence=None, input=None, **inference_kwargs):
```
Comment (gdefe) [inline torch_concepts/nn/modules/high/models/cmr.py:69]: Ideally, it should not be necessary to redefine your forward.

If you need to fill the 'extra' field of the ModelOutput, we have added a method self.default_extra, which can be extended by any model. You can use it to fill anything you need to out.extra

File: torch_concepts/nn/modules/high/models/cmr.py [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+            ordinary = probs[self.task_names]
+            out.params["probs"] = probs[keep]
+            extra = dict(out.extra) if out.extra else {}
+            extra["task_input"] = ordinary
+            extra["input_with_rec"] = rec
+            out.extra = extra
+        return out
+
+    def _build_model(self) -> BayesianNetwork:
+        input_var = EmbeddingVariable("input", distribution=Delta, shape=self.input_size)
```
Comment (gdefe) [inline torch_concepts/nn/modules/high/models/cmr.py:84]: could we standardize this with other models using _input_latent_block? see CBM

File: torch_concepts/nn/modules/low/predictors/rule.py [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+
+class RuleTaskPredictor(BaseConceptLayer):
+    """Compute the ordinary CMR task probability from concepts, selector and roles.
+
+    References:
+        Debot et al. "Interpretable Concept-Based Memory Reasoning", NeurIPS 2024.
+        https://arxiv.org/abs/2407.15527
+    """
+    def forward(self, concepts, selector, roles):
```
Comment (gdefe) [inline torch_concepts/nn/modules/low/predictors/rule.py:46]: in particular, this forward breaks the contract for PyC layers. PyC layers they only take "concepts" and/or "embeddings" as input for the forward.

Is it possible to design CMR task predictor by using a 2-input layer?

File: torch_concepts/nn/modules/high/models/cmr.py [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+
+        roles = EmbeddingVariable("rule_roles", distribution=OneHotCategorical, shape=(len(self.task_names), self.n_rules, n_concepts, 3))
+        roles_cpd = ParametricCPD(roles, parents=[], parametrization={"probs": RuleMemory(len(self.task_names), self.n_rules, n_concepts, self.memory_latent_size, self.memory_decoder_hidden_layers, hard_at_eval=self.hard_roles_at_eval)})
+
+        def aggregate(values):
+            return {"concepts": torch.cat([values[parent] for parent in list(concepts)], dim=-1), "selector": values[selector], "roles": values[roles]}
+
+        tasks = self.build_concept_variables(self.task_names, "tasks")
+        assert len(tasks) == 1, "CMR requires homogeneous binary task variables."
+        task_cpd = ParametricCPD(tasks[0], parents=[*concepts, selector, roles], parametrization={"probs": RuleTaskPredictor(out_concepts=tasks[0].size, in_concepts=n_concepts)}, aggregate=aggregate)
```
Comment (gdefe) [inline torch_concepts/nn/modules/high/models/cmr.py:105]: here you see the consequence of using the three unstandardized params for the forward signature of the task predictor. That is way you need the quite unconventional aggregate function just to match that signature.

Is it possible to design CMR task predictor by using the standard 2-input layer API?

File: torch_concepts/nn/modules/loss.py [hunk 10]
```diff
@@ -597,4 +597,52 @@ def forward(
             mask = torch.isfinite(input)
         if mask.any():
             return self.scale * input[mask].abs().mean()
-        return torch.tensor(0.0, device=input.device)
\ No newline at end of file
+        return torch.tensor(0.0, device=input.device)
+
+
+class CMRBlendedLoss(TypeAwareLoss):
```
Comment (gdefe) [inline torch_concepts/nn/modules/loss.py:603]: This should do the job. But we have added new functionalities to the losses, they should now be more readable and easy to use. 

you can use :
ConceptLoss to route by type, 
ConceptSubset to route by concept name, 
CompositeLoss to weights and sums.

Could you please have a look at whether this helps in designing this loss?  You can use example 2_model/13_composite_loss.py as a reference
